File: topicos/dynamic-programming/group-1/log_n/ log_n.ts
export function logN(x: number): number {
    if (x < 0) return NaN;
    if (x === 0) return Number.NEGATIVE_INFINITY;
    if (x === 1) return 0;

    const z = (x - 1) / (x + 1);
    const powerOdd = z * z;
    
    let serie = 0;
    let base = z;
    let denominator = 1;
    
    const E = 1e-15;

    while (Math.abs(base / denominator) > E) {
        serie += base / denominator;
        base *= powerOdd;
        denominator += 2;
    }

    return 2 * serie;
}

File: topicos/dynamic-programming/group-1/log_n/log_n.rs
fn custom_abs(n: f64) -> {
    if n < 0.0 { -n } else { n }
}

pub fn log_n(x: f64) -> f64 {
    if x < 0.0 { return f64::NAN; }
    if x == 0.0 { return f64::NEGATIVE_INFINITY; }
    if x == 1.0 { return 0.0; }

    let z = (x - 1) / (x + 1);
    let z_square = z * z;

    let mut serie: f64 = 0.0;
    let mut base = z;
    let mut dominator: f64 = 1.0;

    const E: f64 = 1e-15;

    while custom_abs(base/dominator) > E {
        serie += base / dominator;
        base *= z_square;
        dominator += 2.0;
    }

    return 2 * serie
}